plugins/destination/meilisearch/client/read.go

```diff
@@ -45,17 +56,124 @@ func (c *Client) Read(_ context.Context, table *schema.Table, sourceName string,
 	return nil
 }
 
-func unmap(table *schema.Table) func(hit any) ([]any, error) {
-	return func(hit any) ([]any, error) {
-		m, ok := hit.(map[string]any)
-		if !ok {
-			return nil, fmt.Errorf("unsupported format for doc: %T", hit)
+func docToRecord(sc *arrow.Schema, doc map[string]any) (arrow.Record, error) {
+	builder := array.NewRecordBuilder(memory.DefaultAllocator, sc)
+	defer builder.Release()
+
+	for i, fld := range builder.Fields() {
+		val, ok := doc[sc.Field(i).Name]
+		if !ok || val == nil {
+			fld.AppendNull()
+			continue
 		}
 
-		res := make([]any, len(table.Columns))
-		for idx, col := range table.Columns {
-			res[idx] = m[col.Name]
+		switch tBuilder := fld.(type) {
+		case *array.BooleanBuilder:
+			tBuilder.Append(val.(bool))
+		case *array.Int8Builder:
+			tBuilder.Append(int8(val.(float64)))
+		case *array.Int16Builder:
+			tBuilder.Append(int16(val.(float64)))
+		case *array.Int32Builder:
+			tBuilder.Append(int32(val.(float64)))
+		case *array.Int64Builder:
+			tBuilder.Append(int64(val.(float64)))
+		case *array.Uint8Builder:
+			tBuilder.Append(uint8(val.(float64)))
+		case *array.Uint16Builder:
+			tBuilder.Append(uint16(val.(float64)))
+		case *array.Uint32Builder:
+			tBuilder.Append(uint32(val.(float64)))
+		case *array.Uint64Builder:
+			tBuilder.Append(uint64(val.(float64)))
+		case *array.Float32Builder:
+			tBuilder.Append(float32(val.(float64)))
+		case *array.Float64Builder:
+			tBuilder.Append(val.(float64))
+		case *array.StringBuilder:
+			tBuilder.Append(val.(string))
+		case *array.BinaryBuilder:
+			var data []byte
+			val := val.(string)
+			if val != "null" {
+				err := json.Unmarshal([]byte(strconv.Quote(val)), &data)
+				if err != nil {
+					return nil, err
+				}
+			}
+			tBuilder.Append(data)
+		case *array.TimestampBuilder:
+			ts, err := arrow.TimestampFromString(val.(string), arrow.Microsecond)
+			if err != nil {
+				return nil, err
+			}
+			tBuilder.Append(ts)
+		case *types.InetBuilder:
+			_, ipNet, err := net.ParseCIDR(val.(string))
+			if err != nil {
+				return nil, err
+			}
+			tBuilder.Append(*ipNet)
+		case *types.JSONBuilder:
+			tBuilder.Append(val)
+		case *types.MacBuilder:
+			mac, err := net.ParseMAC(val.(string))
+			if err != nil {
+				return nil, err
+			}
+			tBuilder.Append(mac)
+		case *types.UUIDBuilder:
+			uid, err := uuid.Parse(val.(string))
+			if err != nil {
+				return nil, err
+			}
+			tBuilder.Append(uid)
+		case *array.ListBuilder:
```

Comment:
Nested handling can be achieved by recursion, would need a bit of a refactor though

Reply:
Addressed in 6523871